internal/service/kendra/faq.go

```diff
@@ -0,0 +1,445 @@
+	}
+
+	if _, err := waitFaqDeleted(ctx, conn, id, indexId, d.Timeout(schema.TimeoutDelete)); err != nil {
+		return diag.Errorf("waiting for Kendra Faq (%s) delete: %s", d.Id(), err)
+	}
+
+	return nil
+}
+
+func waitFaqCreated(ctx context.Context, conn *kendra.Client, id, indexId string, timeout time.Duration) (*kendra.DescribeFaqOutput, error) {
+
+	stateConf := &resource.StateChangeConf{
+		Pending:                   FaqStatusValues(types.FaqStatusCreating, "PENDING_CREATION"), // API currently returns PENDING_CREATION instead of CREATING
+		Target:                    FaqStatusValues(types.FaqStatusActive),
+		Timeout:                   timeout,
+		Refresh:                   statusFaq(ctx, conn, id, indexId),
+		NotFoundChecks:            20,
+		ContinuousTargetOccurence: 2,
+	}
+
+	outputRaw, err := stateConf.WaitForStateContext(ctx)
+
+	if output, ok := outputRaw.(*kendra.DescribeFaqOutput); ok {
+		if output.Status == types.FaqStatusFailed {
+			tfresource.SetLastError(err, errors.New(aws.ToString(output.ErrorMessage)))
+		}
+		return output, err
+	}
+
+	return nil, err
+}
+
+func waitFaqDeleted(ctx context.Context, conn *kendra.Client, id, indexId string, timeout time.Duration) (*kendra.DescribeFaqOutput, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: FaqStatusValues(types.FaqStatusDeleting, "PENDING_DELETION"), // API currently returns PENDING_DELETION instead of DELETING
+		Target:  []string{},
+		Timeout: timeout,
+		Refresh: statusFaq(ctx, conn, id, indexId),
+	}
+
+	outputRaw, err := stateConf.WaitForStateContext(ctx)
+
+	if output, ok := outputRaw.(*kendra.DescribeFaqOutput); ok {
+		tfresource.SetLastError(err, errors.New(aws.ToString(output.ErrorMessage)))
+
+		return output, err
+	}
+
+	return nil, err
+}
+
+func statusFaq(ctx context.Context, conn *kendra.Client, id, indexId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		output, err := FindFaqByID(ctx, conn, id, indexId)
+
+		if tfresource.NotFound(err) {
+			return nil, "", nil
+		}
+
+		if err != nil {
+			return nil, "", err
+		}
+
+		return output, string(output.Status), nil
+	}
+}
+
+func expandS3Path(tfList []interface{}) *types.S3Path {
+	if len(tfList) == 0 || tfList[0] == nil {
+		return nil
+	}
+
+	tfMap, ok := tfList[0].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	result := &types.S3Path{}
+
+	if v, ok := tfMap["bucket"].(string); ok && v != "" {
+		result.Bucket = aws.String(v)
+	}
+
+	if v, ok := tfMap["key"].(string); ok && v != "" {
+		result.Key = aws.String(v)
+	}
+
+	return result
+}
+
+func flattenS3Path(apiObject *types.S3Path) []interface{} {
+	if apiObject == nil {
+		return nil
+	}
+
+	m := map[string]interface{}{}
+
+	if v := apiObject.Bucket; v != nil {
+		m["bucket"] = aws.ToString(v)
+	}
+
+	if v := apiObject.Key; v != nil {
+		m["key"] = aws.ToString(v)
+	}
+
+	return []interface{}{m}
+}
+
+// Helpers added. Could be generated or somehow use go 1.18 generics?
```

Comment:
yeah i think as we move more steadily towards integrating AWS Go V2, we'll likely have a pattern for this documented